Add ability to extend list of default countries

yuryroot opened the pull request:
Hello, @sshaw!

First, thank you for fast and simple gem!
Second, I want to propose my implementation of feature to extend a default countries list by user data.
I think that such function will be useful.

What do you think about it?

sshaw commented:
Hi, I think this is a nice idea but have some questions/comments:

1. These former countries/republics have legit ISO codes, etc... so why not just add the to the existing list? 
1. One downside to adding them to the existing list is that people using `to_h` and `to_a` for say a country select field will now have what can be considered invalid countries.
1. Why not include a YAML file with a full mapping of deprecated county codes/aliases?

yuryroot commented:
@sshaw, there are my answers on your questions/comments:

1. There are different types of country codes in my private project.
  For example: `US`, `BR`, `RU`, ... and `WO`, `EO`, ...
  In this example `US`, `BR`, `RU` are valid country codes, and `WO`, `EO` are [patent offices](https://en.wikipedia.org/wiki/Patent_office).

  I think that in this case we shouldn't mix different types of countries in the default list.
  I think so, because it will solve only my problem and my custom data are not needed for other people (in particular in methods `to_a` and `to_h`, as you said in the second comment).
  We need a more universal solution.

2. Yes, added countries will be in results of methods `to_a` and `to_h`.
  It is expected behaviour for user who extends the default list...

3. What concrete do you mean?
  Do you mean that we should append a separate YAML config for deprecated countries or append mapping to the existing config?

sshaw commented:
> There are different types of country codes in my private project. For example...
> ... US, BR, RU are valid country codes, and WO, EO are patent offices.

Got it. Makes sense. 

> Do you mean that we can append a separate YAML config for deprecated countries or append 
> mapping to the existing config?

I was thinking that all deprecated codes should be added to `en.yml` with a `deprecated` key  or to a separate `en.yml` file. Then a method could be added say, `NormalizeCountry.use_depricated!`, that would add them to this list.  Though I think this is out of scope for this pull request 😄

sshaw commented on lib/normalize_country.rb:
```diff
@@ -36,12 +36,42 @@ def to_h(key, options = {})
+      Countries.clear
+      remove_instance_variable(:@to) if defined?(@to)
+      remove_instance_variable(:@formats) if defined?(@formats)
+
```
Why not just:  `@to = @formats = nil`?

yuryroot replied:
I wrote just so, because in some cases expression on the right side of the operator `||=` will be executed again until value is `nil`.
A'm agree that `@to = @formats = nil` will also works fine in this case.

sshaw commented on spec/fixtures/extra_countries/en.yml:
```diff
@@ -0,0 +1,7 @@
+SU:
```
I don't think the extra directory is needed. Maybe just call it `spec/fixtures/extra_countries.yml`?

sshaw replied:
Actually is this *and* `EXTRA_COUNTRIES` necessary? `EXTRA_COUNTRIES` can be populated from this or, in the cases where we need a file, it can be `YAML.dump`ed to this. 

What do you think?

sshaw commented on README.rdoc:
```diff
@@ -27,8 +27,13 @@ Convert country names and codes to a standard.
+ NormalizeCountry.convert("United Mexican States") # "MEX"
+
+ # Extend a list of default countries
+ NormalizeCountry.extend_countries(:alpha2 => "SU", :iso_name => "Soviet Union", :short => "USSR")
```
What do you think of `NormalizeCountry.add_countries(...)` (just a thought)?

yuryroot replied:
I think that both variants `add_countries` and `extend_countries` are possible in this case.

We can also have two aliases:
* `add_countries` is more informatively when we want to append new countries.
* `extend_countries` is more informatively when we want to override rules in the existing list.

sshaw replied:
Hi, sorry, forgot about this... 

2 different names for 2 different purposes but there's just one method that provides the same functionality.

sshaw commented on spec/normalize_country_spec.rb:
```diff
@@ -108,6 +114,47 @@
+    let(:file_path) { File.join(File.dirname(__FILE__), "fixtures", "extra_countries", "en.yml") }
+    after { NormalizeCountry.reset! }
+
+    describe "when an array of hashes passed" do
```
Can you use "with" instead of "when" to match [the existing specs that test method arguments](https://github.com/yuryroot/normalize_country/blob/b52b0316d9a336a874ad0d60e817c4b70452d4a3/spec/normalize_country_spec.rb#L98)? E.g., `"with an array of hashes"` no need for "passed".

sshaw commented on spec/normalize_country_spec.rb:
```diff
@@ -108,6 +114,47 @@
+      it_converts_extended_countries
+    end
+
+    describe 'without arguments' do
```
Everything else is in double quotes. Double is preferred.

sshaw commented on spec/test_helper.rb:
```diff
@@ -0,0 +1,21 @@
+        "Soviet Union"   => { :alpha2 => "SU", :iso_name => "Soviet Union", :short => "USSR" },
+        "Czechoslovakia" => { :alpha2 => "CS", :iso_name => "Czechoslovakia" }
+      }.each do |name, (spec, expect)|
+        NormalizeCountry.convert(name, to: spec).must_equal(expect)
```
Can you use a 1.8 style hash here? Everything else uses this style.

sshaw commented on spec/test_helper.rb:
```diff
@@ -0,0 +1,21 @@
+      list = NormalizeCountry.to_a(:iso_name)
+      list.size.must_equal COUNTRY_COUNT + 2
+      ["Soviet Union", "Czechoslovakia"].each { |name| list.must_include(name) }
+    end
```
Why not use `EXTRA_COUNTRIES` here and below? 

```ruby
list.must_include(EXTRA_COUNTRIES.map(&:iso_name))
```

sshaw commented on spec/test_helper.rb:
```diff
@@ -0,0 +1,21 @@
+class Minitest::Spec
```
Why not do something like this:

```ruby
[:IO => File.open(path), :YAML => path, :hash => { ... }].each do |format, value|
  describe "with a #{format}" do 
    before { ... }

    it "extends a list of default countries" do
      # ...
    end

    { "Soviet Union"   => { :alpha2 => "SU", :iso_name => "Soviet Union", :short => "USSR" }, ... }.each do |name, formats|
    formats.each do |format, value|
      # same description as existing specs
      it "normalizes #{name} to #{format}" do
         # ...
      end
   end
 end
```

Though you loose the `*splat` test.

yuryroot commented:
Thanks for you review.
I will do corrections ASAP.

yuryroot commented:
Hello, @sshaw!
I'm sorry for so long absence...

I consider your comments and changed code.
I also found and fixed a problem when method ```extend_countries``` incorrectly processes already existing countries. The method merges such countries now.

Could you please see latest changes?

sshaw commented:
Hi @yuryroot, thanks for the changes. Give me a day or so to have a look. 👓

sshaw commented on lib/normalize_country.rb:
```diff
@@ -36,52 +36,103 @@ def to_h(key, options = {})
+    def add_country(mapping, merge = false)
+      country = Country.new(mapping)
+
+      if merge
```
Wait, what's the purpose of this?

sshaw replied:
Oh, I see. So for all practical this exists to handle the addition of `aliases` or `short`s?

sshaw commented on lib/normalize_country.rb:
```diff
@@ -36,52 +36,103 @@ def to_h(key, options = {})
+
+      if merge
+        existing_country = Countries.values.find do |object|
+          country.mapping.any? { |id, value| value && object[id] == value }
```
Couldn't you just say:
```ruby
country.formats.any? { |format| country[format] == object[format] }
```
No need to add `#mapping`?

sshaw commented:
Closing due to no response. Fell free to open again if you'd like to complete.
